text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1613 @@
+  } else {
+    return setComponentTemplate(
+      precompileTemplate(`<p>Lame</p>`),
+      templateOnly()
+    );
+  }
+}
+```
+
+The problem here is that this requires re-running both the creation of the template-only empty backing instance which has `null` for `this`, and the association between the two. As described above, the template precompilation also happens during the build, which eliminates some but not all of the apparent cost here; but the other parts are needlessly dynamic and expensive.
+
+In this scenario, users can accomplish the same thing by manually hoisting the component definitions to module scope:
+
+```js
+const Cool = <template><p>Cool</p></template>;
+const Lame = <template><p>Lame</p></template>;
+
+function conditionalComponent(predicate) {
+  if (predicate) {
+    return Cool;
+  } else {
+    return Lame;
+  }
+}
+```
+
+(Notice that this is no different than any other concern around other costly operations needlessly happening in a function body repeatedly.)
+
+See further discussion below under [**How We Teach This**](#how-we-teach-this).
+
+
+### Interop
+
+Since all existing components already work using the same low-level primitives the new system uses, strict-mode components using `<template>` can import and invoke components authored in Ember’s loose mode format. Similarly, since loose mode components are resolved to a component which is the default export of a module in the correct/conventional location on disk, components authored in strict mode with `<template>` and exported as the default export in that conventional location will be resolve-able by loose mode components as well.
+
+There are two qualifications to the interop story: a minor one around named exports and a more significant one around non-colocated templates.
+
+
+#### Named exports
+
+However, other components exported as *named* exports will not be available in loose mode: resolution only evaluates default exports. This is a temporary incoherence which will be resolved as the ecosystem migrates to strict mode. As a workaround, users may choose to create reexport-only modules to allow loose-mode access. For example, given a module `app/components/named-only.js` with two named export components:
+
+```js
+export const Greet = <template>
+  <p>Hello, {{@name}}!</p>
+</template>
+
+export const Farewell = <template>
+  <p>Goodbye, {{@name}}!</p>
+</template>
+```
+
+A user wishing to make these available to loose mode could introduce two new modules:
+
+- `app/components/named-only/greet.js`:
+
+    ```js
+    export { Greet as default } from '../named-only';
+    ```
+
+- `app/components/named-only/farewell.js`:
+
+    ```js
+    export { Farewell as default } from '../named-only';
+    ```
+
+Then a loose-mode component could invoke them like `<NamedOnly::Greet @name="Chris">` and `<NamedOnly::Farewell @name="Krycho" />`. Note that this pattern is not at all *necessary* for migration, but may be useful.
+
+
+#### Non-colocated templates
```

Comment:
We should deprecate non-colocated component templates (obviously "until 5.0"), and mention that in this section

Reply:
I’ve added this as a footnote in 42408ab. I only added it as a footnote because (a) I think it’s a separate concern from this RFC—though I’m happy to help push that forward, as I strongly agree; and (b) we still have to introduce the debug warning/error functionality I suggest even if we *do* deprecate them, since the deprecation will not make it clear why they silently fail to render, and many apps silence deprecations until they're ready to tackle them.